### PR TITLE
feat: reduce auth requests

### PR DESCRIPTION
**Pull request by spacewander**

### CODE REVIEW

File: lib/resty/etcd/v3.lua [hunk 1]
```diff
@@ -19,6 +20,7 @@ local decode_json   = cjson.decode
 local encode_json   = cjson.encode
 local encode_base64 = ngx.encode_base64
 local decode_base64 = ngx.decode_base64
+local semaphore = require("ngx.semaphore")
```
Comment (Firstsawyou) [inline lib/resty/etcd/v3.lua:23]: It is better to align with the code format above.

File: lib/resty/etcd/v3.lua [hunk 2]
```diff
@@ -195,34 +205,68 @@ local function choose_endpoint(self)
...
+        if self.jwt_token and now() - self.last_auth_time < 60 * 3 + random(0, 60) then
+            return true, nil
+        end
+
+        if self.last_refresh_jwt_err then
+            utils.log_info("v3 refresh jwt last err: ", self.last_refresh_jwt_err)
+            return nil, self.last_refresh_jwt_err
+        end
+
+        -- something unexpected happened, try again
```
[the remark this answers is not in the document]
Comment (spacewander) [reply lib/resty/etcd/v3.lua:237]: Changed.

File: lib/resty/etcd/v3.lua (hunk 1 above)
Comment (spacewander) [reply lib/resty/etcd/v3.lua]: Changed.

**Review by tokers — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@spacewander do we need to release a new version after we merge this PR?

**Comment by spacewander** [comment]
@membphis 
Let's do it.

**Comment by membphis** [comment]
https://github.com/api7/lua-resty-etcd/pull/104 release version 1.4.3

@spacewander

**Pull request merged**